src/frontend/src/optimizer/plan_node/generic/agg.rs

```diff
@@ -196,33 +196,56 @@ impl<PlanRef: stream::StreamPlanRef> Agg<PlanRef> {
             let mut included_upstream_indices = vec![]; // all upstream indices that are included in the state table
             let mut column_mapping = BTreeMap::new(); // key: upstream col idx, value: table col idx
             let mut table_value_indices = BTreeSet::new(); // table column indices of value columns
-            let mut add_column = |upstream_idx, order_type, is_value| {
-                column_mapping.entry(upstream_idx).or_insert_with(|| {
-                    let table_col_idx =
-                        internal_table_catalog_builder.add_column(&in_fields[upstream_idx]);
-                    if let Some(order_type) = order_type {
-                        internal_table_catalog_builder.add_order_column(table_col_idx, order_type);
+            let mut add_column =
+                |upstream_idx,
+                 order_type,
+                 is_value,
+                 internal_table_catalog_builder: &mut TableCatalogBuilder| {
+                    column_mapping.entry(upstream_idx).or_insert_with(|| {
+                        let table_col_idx =
+                            internal_table_catalog_builder.add_column(&in_fields[upstream_idx]);
+                        if let Some(order_type) = order_type {
+                            internal_table_catalog_builder
+                                .add_order_column(table_col_idx, order_type);
+                        }
+                        included_upstream_indices.push(upstream_idx);
+                        table_col_idx
+                    });
+                    if is_value {
+                        // note that some indices may be added before as group keys which are not
+                        // value
+                        table_value_indices.insert(column_mapping[&upstream_idx]);
                     }
-                    included_upstream_indices.push(upstream_idx);
-                    table_col_idx
-                });
-                if is_value {
-                    // note that some indices may be added before as group keys which are not value
-                    table_value_indices.insert(column_mapping[&upstream_idx]);
-                }
-            };
+                };
 
             for &idx in &self.group_key {
-                add_column(idx, Some(OrderType::ascending()), false);
+                add_column(
+                    idx,
+                    Some(OrderType::ascending()),
+                    false,
+                    &mut internal_table_catalog_builder,
+                );
             }
+            let read_prefix_len_hint = internal_table_catalog_builder.get_current_pk_len();
```

Comment:
Is this the reason we cannot directly capture the builder in the closure?

Reply:
Yes :hot_face: I was just tired and not sure if there could be better method.